### PR TITLE
Docker Support

### PR DESCRIPTION
**Pull request by kireledan**
This pull request adds support for docker provisioning and also some changes to make the addition of other provisioners much simpler.

This PR is in response to #126 .

All functionality with vagrant remains the same - no vagrant related commands have been added.

# Using Molecule with docker

To get started with using molecule with docker, simply initialize using the ```--docker``` flag. This --docker flag has been added to create a simple molecule config that works with docker out of the box. (Comments were removed because they seem extraneous when documentation exists).

Let's initialize a role named foo
``` bash
molecule init foo1 --docker
```

The same directory structure is made, except the molecule.yml file is different.

``` yaml
---
  ansible:
    playbook: playbook.yml
    user: root
    connection: docker

  docker:
    containers:
      - name: foo-01
        ansible_groups:
          - group1
        image: ubuntu
        image_version: latest
      - name: foo-02
        ansible_groups:
          - group2
        image: ubuntu
        image_version: latest
        registry: testhost:5323

```

Docker containers are entered in a similar manner to vagrant instances. The containers contains a list of containers with their name, ansible group, image, and image version.

The ```image``` corresponds to the tag of the repository that the image is pulled from and the ```image_version``` refers to it's version tag. If the image version is a number, it is highly recommended to put it into a string to prevent 0's being taken off. Currently, if the image does not have yum or apt-get get, molecule will not be able to build a compatible image automatically. The ```registry``` can also be added to specify a custom docker registry.

Running ```molecule test``` at this point won't show much. So if we add a quick task of placing a file.

In ```foo1/tasks/main.yml```

``` yaml
---
    - name: Touch a file to ensure it exists
    lineinfile: dest='./sample_file' line="it's alive!" create=ye state=present

```

## Commands

```molecule create``` will bring up the containers if they are not created.
No output indicates that the containers were created already.

``` bash
❯ molecule create                                                                                               
Creating container foo1-01 with base image ubuntu:latest ...  Container created.

Creating container foo1-02 with base image ubuntu:latest ...  Container created.

```

```molecule test``` works as expected and will create fresh containers to provision.

```
❯ molecule test                                                                                                 [10:43:59]
Stopping container foo1-01 ...  Removed container foo1-01.

Stopping container foo1-02 ...  Removed container foo1-02.

Creating container foo1-01 with base image ubuntu:latest ...  Container created.

Creating container foo1-02 with base image ubuntu:latest ...  Container created.


PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [foo1-02]
ok: [foo1-01]

PLAY RECAP *********************************************************************
foo1-01                    : ok=1    changed=0    unreachable=0    failed=0
foo1-02                    : ok=1    changed=0    unreachable=0    failed=0

Idempotence test in progress (can take a few minutes)...
Idempotence test passed.

Skipping tests, docker provisioner does not support integration testing
Stopping container foo1-01 ...  Removed container foo1-01.

Stopping container foo1-02 ...  Removed container foo1-02.

```

The rest of the molecule commands work as expected, except the test will display that there is no integration testing for docker.

# Underlying Changes #

Besides adding Docker support, I made some changes to ensure more provisioners abstraction. The diff should provide enough info on what was changed to add docker support. My goal was to make it less of a hassle to work on each provisioner.

### CODE REVIEW

File: molecule/templates/molecule_docker.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    user: root
+    connection: docker
+
+  docker:
+    containers:
+      - name: {{role}}-01
+        ansible_groups:
+          - group1
+        image: ubuntu
+        image_version: latest
```
Comment (retr0h) [inline molecule/templates/molecule_docker.yml.j2]: Would like to see the ability of pointing to a docker registry to get the image,

Comment (kireledan) [inline molecule/templates/molecule_docker.yml.j2]: Sure thing. I'll push that in my next commit.

**Comment by retr0h** [comment]
> The rest of the molecule commands work as expected, except the test will display that there is no integration testing for docker.

Have you by any chance looked at testinfra's docker connection?  Just curious if it happens to work for us.  http://testinfra.readthedocs.io/en/latest/backends.html#docker

**Comment by kireledan** [comment]
Oh wow. I didn't see that actually. I'll definitely investigate that.

**Comment by lowks** [comment]
Nice! :+1: :+1: :+1:

**Comment by retr0h** [comment]
@kireledan I'd like to see the [docs](https://github.com/metacloud/molecule/blob/master/docs/source/providers.rst) updated breaking out the provisioners and updating docker usage.

Also, would like to know if testinfra will work with this setup.   In a subsequent PR we can make testinfra the default verification runner.

**Comment by kireledan** [comment]
I've merged that PR into my branch because it had a couple of conflicts. Also, Testinfra now works with the docker provisioner.

**Comment by retr0h** [comment]
@kireledan what formatter do you use for the rdoc?  Looks like you were able to batch update spacing.

**Comment by retr0h** [comment]
@kireledan We made a few more changes to #168.  We are hoping to land that first.  Once that is complete we will merge this PR.  You may need to rebase.

**Comment by kireledan** [comment]
I was just editing the rst in Pycharm. Also yes, I will rebase with master after PR 168 is merged in.

**Comment by kireledan** [comment]
rebased.

**Comment by retr0h** [comment]
@kireledan You might need to remove the original rebase work from #168, and try again. Looking at your commit there are changing that I did't change in #168.

https://github.com/metacloud/molecule/pull/165/files#diff-7d3008165b82862fcdec94df0ff555feL426
https://github.com/metacloud/molecule/pull/165/files#diff-77119f14a31a15fad98cf1a427c517b4L199
https://github.com/metacloud/molecule/pull/165/files#diff-b7e1bd03ce89d5601a606dd8e97334d0L1

**Comment by kireledan** [comment]
oh - those are also part of my changes. Unless you mean to make a new PR?

**Comment by retr0h** [comment]
Right, just stating, it looks like you got it backwards.  You rebased against an incomplete PR.

**Comment by kireledan** [comment]
Ahh okay - I see what you mean now. I'll go ahead and rebase against the merged master.

**Comment by kireledan** [comment]
Okay - I've removed the commit where I merged the incomplete PR and rebased against the current master.

**Comment by abrown-sg** [comment]
Wow, this is amazing work! Big 👍 from me. The only thing I'm concerned about is that the YAML in some of the documentation got reformatted to use 4 spaces instead of 2.

**Comment by kireledan** [comment]
Yeah that's my bad - I'll go ahead and fix that.

File: docs/source/usage.rst [hunk 2]
```diff
@@ -86,108 +86,109 @@ Here is a commented example:
...
+		platform:
+		  name: trusty64
+		  box: trusty64
+		  box_url: https://vagrantcloud.com/ubuntu/boxes/trusty64/versions/14.04/providers/virtualbox.box
+		# templates to use when creating files during `molecule init`
+		templates:
+		  molecule: molecule.yml.j2
+		  playbook: playbook.yml.j2
+		  spec_helper: spec_helper.rb.j2
+		  default_spec: default_spec.rb.j2
```
Comment (retr0h) [inline docs/source/usage.rst]: This looks wrong from master.

`spec_helper` and `default_spec` are removed in favor of `test_default`.

File: docs/source/provisioners.rst [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+Vagrant Instance Example
+------------------------
+This is an example of a set of vagrant instance - for information on specifying the platform/
+provider, see :ref:`providers`.
+
+.. code-block:: yaml
+
+    ---
+      instances:
```
Comment (retr0h) [inline docs/source/provisioners.rst]: looks like we need to dedent this 2 spaces.  `---` doesn't line up with `instances`.

File: docs/source/usage.rst [hunk 4]
```diff
@@ -322,6 +322,56 @@ simply be named vagrant-01 instead of vagrant-01-rhel-7.
...
+- apt-get/yum
+- python 2.5+
+- python 2.4 with python-simplejson
+
+Here is an example of a complete ``molecule.yml`` with 2 containers.
+
+.. code-block:: yaml
+
+  ---
+    docker:
```
Comment (retr0h) [inline docs/source/usage.rst]: looks like we need to dedent this.  `---` doesn't line up with `containers`.

File: molecule/commands.py [hunk 5]
```diff
@@ -376,8 +374,14 @@ def execute(self, exit=True):
                              Fore.RESET))
             return None, None
 
+        # # no docker support with testinfra or serverspec
+        # if self.molecule._provisioner.name is 'docker':
+        #     msg = '{}Skipping tests, docker provisioner does not support integration testing'
+        #     print(msg.format(Fore.YELLOW))
+        #     return None, None
+
```
Comment (retr0h) [inline molecule/commands.py]: Can we remove these comments?

**Comment by retr0h** [comment]
`molecule init` creates a file w/ odd spacing.  Needs dedented 2 spaces.

```
---
  ansible:
    playbook: playbook.yml
    user: root
    connection: docker

  docker:
    containers:
      - name: foo-01
        ansible_groups:
          - group1
        image: ubuntu
        image_version: latest
      - name: foo-02
        ansible_groups:
          - group2
        image: ubuntu
        image_version: latest
```

**Comment by retr0h** [comment]
@kireledan I am sure I am doing something wrong.

```
---
ansible:
  config_file: ../../.molecule/ansible.cfg

docker:
  providers:
    - name: docker
      type: docker

  containers:
    - name: foo-01
      ansible_groups:
        - group1
      image: ubuntu
      image_version: latest
    - name: foo-02
      ansible_groups:
        - group2
      image: ubuntu
      image_version: latest
```

```
[jodewey:~/git/ansible-systems/roles/etc_hosts] [venv] test-test-docker(+4/-4)+ ± molecule converge
Traceback (most recent call last):
  File "/Users/jodewey/git/ansible-systems/venv/bin/molecule", line 10, in <module>
    sys.exit(main())
  File "/Users/jodewey/git/molecule/molecule/cli.py", line 68, in main
    CLI().main()
  File "/Users/jodewey/git/molecule/molecule/cli.py", line 63, in main
    c = command_class(command_args, args)
  File "/Users/jodewey/git/molecule/molecule/commands.py", line 80, in __init__
    if self.molecule._provisioner.instances is None:
  File "/Users/jodewey/git/molecule/molecule/provisioners.py", line 247, in instances
    return self.m._config.config['vagrant']['instances']
KeyError: 'instances'
```

**Comment by retr0h** [comment]
@kireledan have you gotten this to work with docker toolbox?

I performed the following:

```
$ docker-machine create --driver virtualbox default
$ docker-machine ls
NAME      ACTIVE   DRIVER       STATE     URL                         SWARM   DOCKER    ERRORS
default   *        virtualbox   Running   tcp://192.168.99.100:2376           v1.11.2
$ eval "$(docker-machine env default)"
```

At this point the docker daemon works and proxies to the vagrant vm.  I was able to run molecule on an example role.

```
$ docker ps
CONTAINER ID        IMAGE                          COMMAND                  CREATED             STATUS              PORTS               NAMES
c3cd7775e498        molecule_local/ubuntu:latest   "bash -c 'sleep infin"   16 minutes ago      Up 16 minutes                           foo-02
b55f30dea0c8        molecule_local/ubuntu:latest   "bash -c 'sleep infin"   16 minutes ago      Up 16 minutes                           foo-01
```

However, I cannot seem to get ansible to work via molecule or on it's own.

```
$ molecule converge

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
fatal: [foo-01]: UNREACHABLE! => {"changed": false, "msg": "SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh", "unreachable": true}
fatal: [foo-02]: UNREACHABLE! => {"changed": false, "msg": "SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh", "unreachable": true}
	to retry, use: --limit @playbook.retry
```

**Comment by retr0h** [comment]
Looks like this works when I add the `connection=docker` to the command line.

$ ANSIBLE_ROLES_PATH=../ ansible-playbook playbook.yml -i .molecule/ansible_inventory --connection=docker

Oddly, doesn't seem to be picking up what is set in `.molecule/ansible_inventory`. 😢

**Comment by retr0h** [comment]
> Oddly, doesn't seem to be picking up what is set in .molecule/ansible_inventory. 😢

I see, this occurs b/c `molecule init --docker` adds the following to molecule.yml

```
ansible:
  connection: docker
```

I wonder if we want to handle this a bit differently.  If one were to have multiple providers, they simply could not switch between the two with `molecule converge --provider=vagrant vs molecule converge --provider=docker`

I'm not entirely sure if we have such a use case tho.

**Comment by kireledan** [comment]
Interesting. I am using docker toolbox actually and have had no issues. I'll try and troubleshoot it in the morning.

**Comment by kireledan** [comment]
Also I cannot think of a use case where you would want to use two different provisioners at the same time. Or also to constantly switch between provisioners.

**Comment by retr0h** [comment]
👍

**Comment by retr0h** [comment]
> Also I cannot think of a use case where you would want to use two different provisioners at the same time. Or also to constantly switch between provisioners.

Yeah, most of what we have today is in Vagrant, and we will be moving our CI to docker (thanks to your wonderful PR).  However, we will likely want to retain both use cases for some time.

* Do you think you could update the docs with examples of how to use molecule on travis ci?  I think thats one of the biggest wins with this molecule update.  Curious if you have tried it.
* How do you feel about making connection switch depending on the provider?

**Comment by kireledan** [comment]
Glad to hear it :). Also, Im unable to reproduce the errors you are getting. Several of my coworkers have run molecule with docker-machine and have not had any issues. I plan on moving some ansible connection parameters to the provisioner level - that might fix the issues you are having and will allow for multiple provisioners at once (and also make molecule easier to configure for others)

I will also investigate using molecule on travis.ci as I have not used it on there myself.

**Comment by retr0h** [comment]
> Glad to hear it :). Also, Im unable to reproduce the errors you are getting. Several of my coworkers have run molecule with docker-machine and have not had any issues. I plan on moving some ansible connection parameters to the provisioner level - that might fix the issues you are having and will allow for multiple provisioners at once (and also make molecule easier to configure for others)

Yeah, I am good.  My problem was the `connection: ansible` missing in molecule.yml. :)

**Comment by kireledan** [comment]
Ahh okay. I just mis-read your comments. Oops.

**Comment by retr0h** [comment]
> Ahh okay. I just mis-read your comments. Oops.

Not sure if the docs should be updated to reflect that.  Was kinda thinking `connection` might be conditional based on provider.

**Comment by retr0h** [comment]
Needs a rebase.

**Comment by kireledan** [comment]
I've added a travis example and also my branch has updated the travis file to be able to test the docker provisioner.

**Comment by retr0h** [comment]
👍

**Comment by abrown-sg** [comment]
💯 Beers on me too!

**Comment by abrown-sg** [comment]
:boom:

**Pull request merged**